util/gitutil/git_protocol.go

```diff
@@ -1,46 +1,66 @@
 package gitutil
 
 import (
+	"net/url"
+	"regexp"
 	"strings"
 
 	"github.com/moby/buildkit/util/sshutil"
 )
 
 const (
-	HTTPProtocol = iota + 1
-	HTTPSProtocol
-	SSHProtocol
-	GitProtocol
-	UnknownProtocol
+	HTTPProtocol    = "http"
+	HTTPSProtocol   = "https"
+	SSHProtocol     = "ssh"
+	GitProtocol     = "git"
+	UnknownProtocol = "unknown"
 )
 
-// ParseProtocol parses a git URL and returns the remote url and protocol type
-func ParseProtocol(remote string) (string, int) {
-	prefixes := map[string]int{
-		"http://":  HTTPProtocol,
-		"https://": HTTPSProtocol,
-		"git://":   GitProtocol,
-		"ssh://":   SSHProtocol,
-	}
-	protocolType := UnknownProtocol
-	for prefix, potentialType := range prefixes {
-		if strings.HasPrefix(remote, prefix) {
-			remote = strings.TrimPrefix(remote, prefix)
-			protocolType = potentialType
+var protos = map[string]struct{}{
+	"http":  {},
+	"https": {},
+	"git":   {},
+	"ssh":   {},
+}
+
+var hasProtoRegexp = regexp.MustCompile(`^[a-z0-9]+://`)
+
+// ParseProtocol parses a git URL and returns the protocol type, the remote,
+// and the path.
+//
+// ParseProtocol understands implicit ssh URLs such as "git@host:repo", and
+// returns the same response as if the URL were "ssh://git@host/repo".
+func ParseProtocol(remote string) (string, string, string) {
+	if hasProtoRegexp.MatchString(remote) {
+		u, err := url.Parse(remote)
+		if err != nil {
+			return remote, "", UnknownProtocol
 		}
-	}
 
-	if protocolType == UnknownProtocol && sshutil.IsImplicitSSHTransport(remote) {
-		protocolType = SSHProtocol
-	}
+		proto := u.Scheme
+		_, ok := protos[proto]
+		if !ok {
+			proto = UnknownProtocol
```

Comment:
Maybe return an error?

Reply:
Done. I've also reworked the return values so that the following works:

```go
	proto, remote, path, err := gitutil.ParseProtocol(url)
	if err != nil && proto == gitutil.UnknownProtocol {
		url = "https://" + url
```

Unknown protocol now means that the proto regexp did not match, so we can safely prepend "https://" and re-attempt parsing (instead of always doing so, which could result in us attempting to parse weird things like `https://foo://example.com/moby/buildkit.git`).